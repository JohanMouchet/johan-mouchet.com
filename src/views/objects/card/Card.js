// @flow

import * as React from "react";
import cx from "classnames";
import "./card.scss";

type Props = {
  banner: ?string,
  bannerUrl: ?string,
  header: ?string,
  body: ?string,
  footer: ?string,
  metadata: {
    contextClass: ?string,
    additionalClasses: ?Array<string>,
    modifierClasses: ?Array<string>
  }
};

const Card = (props: Props) => {
  const { prop1, prop2 } = props;

  const CardClass = cx(
    "o-card",
    props.metadata.contextClass,
    props.metadata.modifierClasses,
    props.metadata.additionalClasses,
    {
      "Card--prop1": prop1
    }
  );

  const Tag = !props.header ? "header" : "div";

  return (
    <section className={CardClass}>
      {(props.banner || props.bannerUrl) && (
        <Tag
          className="o-card__banner"
          style={
            props.bannerUrl && `backgroundImage: url('${props.bannerUrl}')`
          }
        >
          {props.banner === true ? "" : props.banner}
        </Tag>
      )}
      {/* TODO: <Card><Header>Sub comp</Header></Card> */}
      {props.header && (
        <header className="o-card__header">{props.header | raw}</header>
      )}
      {props.body && <div className="o-card__body">{props.body | raw}</div>}
      {props.footer && (
        <footer className="o-card__footer">{props.footer | raw}</footer>
      )}
    </section>
  );
};

Card.defaultProps = {
  prop1: 2
};

export default Card;
