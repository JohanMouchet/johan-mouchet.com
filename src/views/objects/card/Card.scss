@import "assets/styles/settings/_settings";
@import "assets/styles/tools/_tools";

.o-card {
  $root: &;
  $v-padding: t-vertical-rhythm(0.75) !default;
  $h-padding: $v-padding * 2;
  $scale-factor: 1.5 !default;

  /* Group */
  &-group {
    margin-bottom: $grid-gutter;

    .grid--lgGutter-\@xs > & {
      // TODO: find a way to import single vendor
      // margin-bottom: $grid-gutter-lg;
    }
  }

  /* Card */
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  border: 1px solid $s-color-gray;
  background-color: $s-color-white;
  transition: box-shadow 0.2s, transform 0.2s;

  /* Components */
  &__banner,
  &__header,
  &__body,
  &__footer {
    flex: 0 1 auto;

    > *:first-child {
      margin-top: 0;
    }

    > *:last-child {
      margin-bottom: 0;
    }
  }

  /* Banner */
  &__banner {
    min-height: t-vertical-rhythm(7);
    padding: $v-padding $h-padding;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    position: relative;
    background-color: $s-color-gray-lighter;
    background-position: center center;
    background-repeat: no-repeat;
    background-size: cover;
    border-bottom: 1px solid;
    border-bottom-color: inherit;
    border-top-left-radius: inherit;
    border-top-right-radius: inherit;

    /* stylelint-disable-next-line no-descending-specificity */
    * {
      color: $s-color-white;
      text-shadow: -1px 1px 1px rgba($s-color-black, 0.3);
      z-index: map-get($s-layers, card-banner);
    }
  }

  /* Header */
  &__header {
    padding: $v-padding $h-padding;
    border-bottom: 1px solid;
    border-bottom-color: inherit;
  }

  /* Body */
  &__body {
    padding: $v-padding $h-padding;
    flex-grow: 1;
  }

  /* Footer */
  &__footer {
    padding: $v-padding $h-padding;
    border-top: 1px solid;
    border-top-color: inherit;
  }

  /** Hover / Forward **/
  &--hover:hover,
  &--hover:focus,
  &--forward {
    transform: translate(-4px, -4px);
  }

  /** Active **/
  &--active:active {
    box-shadow: none;
    transform: none;
  }

  /** Spacing **/

  /*** Small ***/
  &--sm & {
    &__banner,
    &__header,
    &__body,
    &__footer {
      padding: ($v-padding / $scale-factor) ($h-padding / $scale-factor);
    }
  }

  /*** Large ***/
  &--lg & {
    &__banner,
    &__header,
    &__body,
    &__footer {
      padding: $v-padding $h-padding;
    }

    @media (min-width: map-get($s-breakpoints, sm)) {
      &__banner,
      &__header,
      &__body,
      &__footer {
        padding: ($v-padding * $scale-factor) ($h-padding * $scale-factor);
      }
    }
  }

  /** Secondary **/
  &--secondary {
    border-color: $s-color-secondary;

    #{$root}__hr {
      border-color: $s-color-secondary;
    }
  }

  /** Primary **/
  &--primary {
    border-color: $s-color-primary;

    #{$root}__hr {
      border-color: $s-color-primary;
    }
  }

  /** Banner gradient **/
  &--banner-gradient & {
    &__banner::after {
      content: "";
      background-image: linear-gradient(
        to top,
        rgba($s-color-black, 0.3) 0%,
        transparent 50%
      );
      height: 100%;
      width: 100%;
      position: absolute;
      bottom: 0;
      left: 0;
      border-top-left-radius: inherit;
      border-top-right-radius: inherit;
    }
  }
}
