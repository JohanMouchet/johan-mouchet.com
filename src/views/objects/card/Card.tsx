import React from "react";
import cx from "classnames";
import styles from "./Card.module.scss";

type Props = {
  bannerUrl?: string;
  bannerGradient?: boolean;
  header?: React.ReactNode;
  footer?: React.ReactNode;
  variant?: "secondary" | "primary" | "hover" | "hover-active" | "forward";
  size?: "sm" | "lg";
  className?: string | string[] | { [key: string]: boolean };
  children?: React.ReactNode;
};

const Card: React.FC<Props> = ({
  bannerUrl,
  bannerGradient,
  header,
  children,
  footer,
  variant,
  size,
  className,
}) => {
  const classNames = cx(
    styles["o-card"],
    bannerGradient && styles[`o-card--banner-gradient`],
    variant && styles[`o-card--${variant}`],
    size && styles[`o-card--${size}`],
    className
  );

  return (
    <section className={classNames}>
      {bannerUrl && (
        <div
          className={styles["o-card__banner"]}
          style={
            bannerUrl ? { backgroundImage: `url(${bannerUrl})` } : undefined
          }
        ></div>
      )}
      {header && <header className={styles["o-card__header"]}>{header}</header>}
      {children && <div className={styles["o-card__body"]}>{children}</div>}
      {footer && <footer className={styles["o-card__footer"]}>{footer}</footer>}
    </section>
  );
};

export default Card;
