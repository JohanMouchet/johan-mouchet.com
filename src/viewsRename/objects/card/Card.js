// @flow

import * as React from "react";
import cx from "classnames";
import "./card.scss";

type Props = {
  banner?: string,
  bannerUrl?: string,
  header?: Node,
  body?: Node,
  footer?: Node,
  metadata?: {
    contextClass?: string,
    additionalClasses?: Array<string>,
    modifierClasses?: Array<string>
  }
};

const Card = (props: Props) => {
  const { prop1, metadata = {} } = props;

  const CardClass = cx(
    "o-card",
    metadata.contextClass,
    metadata.modifierClasses,
    metadata.additionalClasses,
    {
      "Card--prop1": prop1
    }
  );

  const Tag = !props.header ? "header" : "div";

  return (
    <section className={CardClass}>
      {(props.banner || props.bannerUrl) && (
        <Tag
          className="o-card__banner"
          style={
            props.bannerUrl && `backgroundImage: url('${props.bannerUrl}')`
          }
        >
          {props.banner === true ? "" : props.banner}
        </Tag>
      )}
      {/* TODO: <Card><Header>Sub comp</Header></Card> */}
      {props.header && (
        <header className="o-card__header">{props.header}</header>
      )}
      {props.body && <div className="o-card__body">{props.body}</div>}
      {props.footer && (
        <footer className="o-card__footer">{props.footer}</footer>
      )}
    </section>
  );
};

Card.defaultProps = {
  prop1: 2
};

export default Card;
