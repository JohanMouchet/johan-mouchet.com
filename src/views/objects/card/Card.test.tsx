import { render } from "@testing-library/react";
import { Card } from "./Card";

it("Renders <Card /> without crashing", () => {
  expect(
    render(
      <Card
        bannerUrl="https://via.placeholder.com/400x168"
        bannerGradient
        header="Header"
        footer="Footer"
        variant="secondary"
        size="sm"
      >
        children
      </Card>
    ).container.firstChild
  ).toMatchSnapshot();
});
