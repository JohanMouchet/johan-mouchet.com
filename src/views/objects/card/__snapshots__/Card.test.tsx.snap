// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Card /> without crashing 1`] = `
<section
  class="o-card o-card--banner-gradient o-card--secondary o-card--sm"
>
  <div
    class="o-card__banner"
    style="background-image: url(https://via.placeholder.com/400x168);"
  />
  <header
    class="o-card__header"
  >
    Header
  </header>
  <div
    class="o-card__body"
  >
    children
  </div>
  <footer
    class="o-card__footer"
  >
    Footer
  </footer>
</section>
`;
